### ISSUE TITLE
`CLUSTER_WIDE` transactions routinely exhibit Fractured Read

### ISSUE THREAD

**Issue by aphyr** [ravendb/ravendb#17928]
Building on #17927, in which `SINGLE_NODE` transactions exhibit lost updates on single-node clusters, I've also explored transactions which use `sessionOptions.setTransactionMode(CLUSTER_WIDE)`. Per RavenDB's [cluster-wide transactions documentation](https://ravendb.net/docs/article-page/6.0/csharp/server/clustering/cluster-transactions), "Concurrent cluster-wide transactions are guaranteed to appear as if they are run one at a time (serializable isolation level)." However, it appears that CLUSTER_WIDE transactions routinely allow G-single, G2-item, and Fractured Read. even in a single-node cluster.  These anomalies are prohibited by Serializability, Repeatable Read, Snapshot Isolation, Forward Consistent View, Consistent View, Parallel Snapshot Isolation, Prefix, Causal, Update Atomic, and Read Atomic.

For example, consider [20231229T105001.666-0600.zip](https://github.com/ravendb/ravendb/files/13795009/20231229T105001.666-0600.zip), in which two threads perform entirely `CLUSTER_WIDE` transactions on a fresh single-node instance of RavenDB 6.0.2. In five seconds we performed 4693 transactions, and observed 8 instances of G-single, 63 instances of G2-item, and 9 instances of G2-item-realtime.

Here's one instance of G-single & Fractured Read from that test run, involving two transactions. The bottom transaction appends 15 to key 103, then appends 9 to key 104. The top transaction reads key 103 and observes `[1 2 4 5 7 10 12]`--failing to see the bottom transaction's append of 15. It goes on to read key 104, and sees `[1 2 3 4 7 9]`--including the bottom transaction's append of 9.

![Screenshot from 2023-12-29 11-06-24](https://github.com/ravendb/ravendb/assets/3748/00b52a14-f5ef-4355-bfc1-7cfccc67b0a0)

In short, the top transaction observed some, but not all, of the bottom transaction's writes. Is there... maybe something else you have to do, on top of turning on `CLUSTER_WIDE` transactions, to obtain Serializability?

**Comment by ayende** [comment]
Hi,
Thanks for the detailed report and in general for taking a look at RavenDB. 

If I understand this properly, you are doing something similar to:

````
TX 1:
 - read A
 - read B
 - if A.Set is not true:
 -- modify B - Set = true
- SaveCchanges()

TX 2:
 - read A
 - read B
 - if B.Set is not true:
 -- modify A - Set = true
- SaveCchanges()
````

And see that both of those are proceeding, right? (Sorry, I couldn't find the code for executing this, and I'm afraid that I'm not familiar enough with Jepsen to read through the linked file).

The client API send to the server only the _modified_ documents in the session. In this case, the transactions of both of those completed with just one of those, and could thus observe the invalid state. 

Reminder: A transaction in RavenDB is _a_ request - so TX1 and TX2 above aren't actually single transactions, instead, each of them represent 3 independent transactions. 
This is because the common use case and what we are optimizing RavenDB for is web application, where a read is sent to the suer, and a separate request will be done for the read, see the fuller explanation in https://github.com/ravendb/ravendb/issues/17927.

Could you point me to the RavenDB code that you are running in this context? 

I'm speaking from memory here, haven't fully checked the code, but IIRC, we'll only apply changes to items that were modified in cluster wide transactions unless you explicitly ask for it.

**Comment by aphyr** [comment]
No, there's no conditionals here. It's effectively just:

```
TX 1:
- read A
- read B
- SaveChanges()

TX 2:
- read A
- Modify the in-memory representation of A
- store A
- read B
- Modify the in-memory representation of B
- store B
- SaveChanges()
```

I linked to the RavenDB transaction code in the other issue, but here it is again, just in case: https://github.com/jepsen-io/ravendb/blob/ff829611ae4d2dd4d21d14bd6eebc75e8924ad5b/src/jepsen/ravendb/append.clj#L26-L40

> A transaction in RavenDB is a request - so TX1 and TX2 above aren't actually single transactions, instead, each of them represent 3 independent transactions.

As I mentioned in [the other issue](https://github.com/ravendb/ravendb/issues/17927#issuecomment-1873023835), this is... maybe the best word is alarming? The generally understood purpose of transactions is to make multiple micro-operations--typically, a series of reads and writes interspersed with computation--execute as a single, atomic, isolated unit. RavenDB's documentation repeatedly insists a session represents a transaction. You've [directly compared](https://www.youtube.com/watch?v=5ZXBR3croMA&t=23m40s) a session to a MongoDB transaction. RavenDB's concurrency control features, like "optimistic concurrency" and "cluster-wide transactions", have a kind of version tracking that's clearly trying to provide *some* transactional isolation spanning across reads and writes in a session. That's... how a transaction would work, in basically any other database. Are you... are you sure about this?

**Comment by ayende** [comment]
Thanks, that is much clearer. Thanks for the interest in RavenDB, and let's see if I can clear things out on my end as well.

A transaction in RavenDB (read, write or mixed) is limited to a _single_ HTTP request. We don't support a transaction over more than a single HTTP request. The [documentation talks about modifications](https://ravendb.net/docs/article-page/6.0/csharp/client-api/faq/transaction-support#acid-for-document-operations), but it applies to reads as well. 

The code below is actually issuing two _separate_ requests (and hence two separate transactions) in the scope of a single session. (Note, there is a concept of a business transaction, which I'm answering on https://github.com/ravendb/ravendb/issues/17927, if it isn't a problem, let's have a single thread of discussion there).

In other words:

```
TX 1.1:
- read A
TX 1.2:
- read B

 # noop, does nothing
- SaveChanges()
```

There is no expectation in RavenDB that this would be transaction at all.

[A little bit later in the video you referenced](https://youtu.be/5ZXBR3croMA?t=1775) I'm talking about how this actually works in RavenDB. Note that the API call that I'm using there is loading _two_ documents at the same time.

That is an important aspect, because the call to `await session.load<Account>(from,to)` is a single request to the server, which will read both of the documents in a single transaction. 

Both account documents are then mutated and another document is created, all of this is done without any involvement on the server. Indeed, there isn't any server-side reference to the session.

The next part of the operation completes with a call to `await session.saveChanges()` - at this point, the server gets all the changes on the session and run through them. Because the session is setup with `useOptimisticConcurrency = true`, it will also ensure that the current version of the documents matches their version at the time we read them.

**Comment by arekpalinski** [comment]
We have updated our documentation and provided more clarifications when talking about transactions in RavenDB and the Client API Session. 

The actual changes we made can be reviewed here: https://github.com/ravendb/docs/pull/1766.
